File: RayTracerChallenge/Sphere.cs
//-----------------------------------------------------------------------
// <copyright file="Sphere.cs" company="Andy Young">
//     Copyright (c) Andy Young. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
namespace OnSubmit.RayTracerChallenge
{
    using System;

    /// <summary>
    /// Represents a sphere.
    /// </summary>
    public class Sphere : Shape
    {
        /// <summary>
        /// Gets the intersection distances along the ray with the sphere.
        /// </summary>
        /// <param name="ray">The ray.</param>
        /// <returns>The intersections along the ray.</returns>
        public override Intersections GetIntersectionsWithImpl(Ray ray)
        {
            Tuple4D sphereToRay = ray.Origin - Tuple4D.CreatePoint(0, 0, 0);

            double a = ray.Direction.GetDotProductWith(ray.Direction);
            double b = 2 * ray.Direction.GetDotProductWith(sphereToRay);
            double c = sphereToRay.GetDotProductWith(sphereToRay) - 1;
            double discriminant = (b * b) - (4 * a * c);

            if (discriminant >= 0)
            {
                double sqrtDiscriminant = Math.Sqrt(discriminant);
                double twoA = 2 * a;

                return new Intersections(new Intersection[2]
                {
                    new Intersection((-b - sqrtDiscriminant) / twoA, this),
                    new Intersection((-b + sqrtDiscriminant) / twoA, this),
                });
            }

            return new Intersections();
        }

        /// <summary>
        /// Gets the surface normal vector at the given point.
        /// </summary>
        /// <param name="point">The point.</param>
        /// <returns>The surface normal vector.</returns>
        public override Tuple4D GetNormalAtPointImpl(Tuple4D point)
        {
            return point - Tuple4D.CreatePoint(0, 0, 0);
        }
    }
}

File: TypeScript/src/ts/shapes/Sphere.ts
import Intersection from "ts/Intersection";
import Intersections from "ts/Intersections";
import Material from "ts/Material";
import Point from "ts/Point";
import Ray from "ts/Ray";
import Vector from "ts/Vector";
import Shape from "./Shape";

export default class Sphere extends Shape {
  override type = "Sphere";

  readonly origin: Point;
  readonly radius: number;

  constructor(material: Material = new Material()) {
    super(material);

    this.origin = new Point(0, 0, 0);
    this.radius = 1;
  }

  override compare = (shape: Shape): boolean => {
    if (!(shape instanceof Sphere)) {
      return false;
    }

    const sphere = shape as Sphere;
    if (!this.origin.compare(sphere.origin)) {
      return false;
    }

    if (!this.radius.compare(sphere.radius)) {
      return false;
    }

    return true;
  };

  override getNormalAtImpl = (point: Point): Vector => point.subtractPoint(this.origin);

  override getIntersectionsWithImpl = (ray: Ray): Intersections => {
    const sphereToRay = ray.origin.subtractPoint(Point.origin);

    const a = ray.direction.dot(ray.direction);
    const b = 2 * ray.direction.dot(sphereToRay);
    const c = sphereToRay.dot(sphereToRay) - 1;
    const discriminant = b * b - 4 * a * c;

    if (discriminant >= 0) {
      const sqrtDiscriminant = Math.sqrt(discriminant);
      const twoA = 2 * a;

      return new Intersections(
        new Intersection((-b - sqrtDiscriminant) / twoA, this),
        new Intersection((-b + sqrtDiscriminant) / twoA, this)
      );
    }

    return new Intersections();
  };
}
